### PR TITLE
TYP: fix typing errors for mypy==0.790, bump mypy version

### LINKED ISSUE #37104: TYP: typing errors for mypy==0.790
```
mypy pandas
pandas/core/computation/parsing.py:41: error: unused 'type: ignore' comment
pandas/io/common.py:452: error: Unsupported operand types for + ("List[Optional[str]]" and "List[str]")  [operator]
pandas/core/aggregation.py:483: error: Value of type variable "_LT" of "sorted" cannot be "Optional[Hashable]"  [type-var]
pandas/core/aggregation.py:606: error: Value of type variable "_LT" of "sorted" cannot be "Optional[Hashable]"  [type-var]
pandas/core/groupby/ops.py:270: error: Incompatible types in assignment (expression has type "Index", variable has type "Iterator[Any]")  [assignment]
pandas/core/groupby/ops.py:270: note: 'Index' is missing following 'Iterator' protocol member:
pandas/core/groupby/ops.py:270: note:     __next__
Found 5 errors in 4 files (checked 1057 source files)
```

### PR DESCRIPTION
**Pull request by fangchenli**
- [x] closes  #37104
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

**Comment by jreback** [comment]
cc @simonjayhawkins

File: pandas/core/aggregation.py [hunk 1]
```diff
@@ -480,7 +480,7 @@ def transform_dict_like(
 
     if obj.ndim != 1:
         # Check for missing columns on a frame
-        cols = sorted(set(func.keys()) - set(obj.columns))
+        cols = sorted(set(func.keys()) - set(obj.columns))  # type: ignore[type-var]
```
Comment (simonjayhawkins) [inline pandas/core/aggregation.py:483]: when adding ignores, can you add the mypy error message as a comment.

This is apparently not a false positive, see https://github.com/python/mypy/issues/9582, so should probably fix and not ignore.

File: pandas/core/aggregation.py [hunk 2]
```diff
@@ -603,7 +603,9 @@ def aggregate(obj, arg: AggFuncType, *args, **kwargs):
             if isinstance(selected_obj, ABCDataFrame) and len(
                 selected_obj.columns.intersection(keys)
             ) != len(keys):
-                cols = sorted(set(keys) - set(selected_obj.columns.intersection(keys)))
+                cols = sorted(
+                    set(keys) - set(selected_obj.columns.intersection(keys))
+                )  # type: ignore[type-var]
```
Comment (simonjayhawkins) [inline pandas/core/aggregation.py:608]: same

File: pandas/io/common.py [hunk 3]
```diff
@@ -448,9 +448,11 @@ def infer_compression(
     if compression in _compression_to_extension:
         return compression
 
-    msg = f"Unrecognized compression type: {compression}"
-    valid = ["infer", None] + sorted(_compression_to_extension)
-    msg += f"\nValid compression types are {valid}"
+    infer = {"infer": None}
+    msg = (
+        f"Unrecognized compression type: {compression}\n"
+        f"Valid compression types are {_compression_to_extension} and {infer}"
```
Comment (simonjayhawkins) [inline pandas/io/common.py:454]: This is changing the error message? Do we have a test that checks the message?

we have a few cases where using list literals produces this mypy error. This one is probably OK to ignore.

**Review by simonjayhawkins — COMMENTED** [review]
Thanks @fangchenli for the PR. generally lgtm.

File: pandas/core/aggregation.py (hunk 1 above)
Comment (simonjayhawkins) [reply pandas/core/aggregation.py:483]: can you do this.

Comment (fangchenli) [reply pandas/core/aggregation.py:483]: I ran into the problem with Label. I'll close this for now.

File: pandas/io/common.py (hunk 3 above)
Comment (fangchenli) [reply pandas/io/common.py]: No, we don't have a test for the message.

File: pandas/core/aggregation.py (hunk 1 above)
Comment (simonjayhawkins) [reply pandas/core/aggregation.py]: fixed in #39028

File: pandas/core/aggregation.py (hunk 2 above)
Comment (simonjayhawkins) [reply pandas/core/aggregation.py]: also fixed in #39028

File: pandas/io/common.py [hunk 4]
```diff
@@ -179,7 +179,8 @@ def stringify_path(
         # this function with convert_file_like=True to infer the compression.
         return cast(FileOrBuffer[AnyStr], filepath_or_buffer)
 
-    if isinstance(filepath_or_buffer, os.PathLike):
+    # Only @runtime_checkable protocols can be used with instance and class checks
+    if isinstance(filepath_or_buffer, os.PathLike):  # type: ignore[misc]
```
Comment (simonjayhawkins) [inline pandas/io/common.py:183]: see https://github.com/pandas-dev/pandas/issues/37104#issuecomment-756673310

File: pandas/io/common.py [hunk 5]
```diff
@@ -487,9 +488,15 @@ def infer_compression(
...
         return compression
 
-    msg = f"Unrecognized compression type: {compression}"
-    valid = ["infer", None] + sorted(_compression_to_extension)
-    msg += f"\nValid compression types are {valid}"
+    # https://github.com/python/mypy/issues/5492
+    # Unsupported operand types for + ("List[Optional[str]]" and "List[str]")
+    valid = ["infer", None] + sorted(
+        _compression_to_extension
+    )  # type: ignore[operator]
```
Comment (simonjayhawkins) [inline pandas/io/common.py:495]: see https://github.com/pandas-dev/pandas/issues/37104#issuecomment-709921704

**Review by simonjayhawkins — APPROVED** [review]
Thanks @fangchenli lgtm pending green

**Review by simonjayhawkins — CHANGES_REQUESTED** [review]
I think we also added the whatnew dependencies on the last bump

**Comment by jreback** [comment]
looks ok to me, merge master just to confirm.

**Comment by simonjayhawkins** [comment]
Thanks @fangchenli

**Comment by simonjayhawkins** [comment]
> merge master just to confirm.

np-dev failures seen on other PRs

**Pull request merged**